### PR TITLE
Select appropriate VB runtime by default

### LINKED ISSUE #2793: Result of chr() cannot be used as initial value of a Const in .Net Core 3 Preview 1
Version: 3.0.100-preview-009812
Commit: e3abf6e935

Runtime Environment:
OS Name: Windows
OS Version: 10.0.17763
OS Platform: Windows
RID: win10-x64
Base Path: C:\Program Files\dotnet\sdk\3.0.100-preview-009812\

Host (useful for support):
Version: 3.0.0-preview-27122-01
Commit: 00c5c8bc40

Have you experienced this same bug with .NET Framework?: No

In the .Net Framework using VB.Net you can compile this Line:

``` vb
Private Const TabChar As Char = Chr(9)
``` 

In .Net Core 3 Preview 1 it failes with:
error BC30059: Constant expression is required.

### LINKED ISSUE #3387: VB.NET "My" not working as expected?
@cston and @KathleenDollard,

Here is an isolated example running on the most recent bits as far as I can tell (at least with regards to the Microsoft.VisualBasic side of things).

Module Program
  Sub Main(args As String())
    Console.WriteLine(My.Computer.Name)
    Console.WriteLine((New Microsoft.VisualBasic.Devices.ServerComputer).Name)
    Console.WriteLine(System.Environment.MachineName)
  End Sub
End Module

The first line doesn’t work in .NET Core 3.0; however, the second and third lines do.  Playing with the same code in the full .NET Framework, all three lines work and when you jump to the reference for My.Computer.Name, it goes to Microsoft.VisualBasic.Devices.ServerComputer.Name.  I then jumped back over to the .NET Core source tree and that property simply returns System.Environment.MachineName.  One thing I find interesting about this is it’s not a 1:1 map… how does My.Computer.Name map to a property inside of an instantiated instance of Microsoft.VisualBasic.Devices.ServerComputer?  Some “magic” black-box stuff there?

So am I just missing something as far as how to "light up" the My side of things when working with a .NET Core 3.0 console project?  Or is there some additional work that still has to be done beyond the work that has taken place within the Microsoft.VisualBasic namespace?

(BTW, @cston great work thus far!)

Thanks.

Cory Smith

### PR DESCRIPTION
**Pull request by nguerrera**
Fix #3387 
Fix #2793 

This comment in the code describes and documents the approach:

```
By default, VBRuntime is chosen as follows:

    .NET Framework:
      Same as classic vbproj. $(FrameworkPathOverride) is passed  as
      /sdkpath to compiler and Microsoft.VisualBasic.dll will be used
      from there even if there is no explicit reference to it. This is
      accomplished by leaving $(VBRuntime) unset.

    .NET Core 3.0+:
      Locate Microsoft.VisualBasic among references and pass it as
      /vbruntime to compiler.

    .NET Core < 3.0 or .NET Standard:
      Use embedded VB runtime as there is no Microsoft.VisualBasic.dll
      available that is complete enough to serve as /vbruntime.

   These defaults can be overridden as follows:

    $(UseReferencedVBRuntime) == true:
      Use referenced Microsoft.VisualBasic regardless of TFM.

      This is an escape hatch for a custom Microsoft.VisualBasic reference
      that can work as /vbruntime on a target framework other than .NET Core 3.0+

    $(VBRuntime) != '':
      Pass it through to compiler unchanged.
```

### CODE REVIEW

**Review by livarcocc — APPROVED** [review]

File: src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToUseVB.cs [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+            var outputDirectory = buildCommand.GetOutputDirectory(targetFramework);
+            var actualVBRuntime = GetVBRuntime(buildCommand.GetValues().FirstOrDefault());
+            File.Delete(outputDirectory.File("VBRuntimeValues.txt").FullName);
+
+            outputDirectory.Should().OnlyHaveFiles(expectedOutputFiles);
+            actualVBRuntime.Should().Be(expectedVBRuntime);
+        }
+
+        private static(VBRuntime, string[]) GetExpectedOutputs(string targetFramework)
```
Comment (nguerrera) [inline src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToUseVB.cs]: self-nit: missing space

```suggestion
        private static (VBRuntime, string[]) GetExpectedOutputs(string targetFramework)
```

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.VisualBasic.targets [hunk 2]
```diff
@@ -60,4 +60,52 @@ Copyright (c) .NET Foundation. All rights reserved.
...
+
+  <PropertyGroup Condition="'$(UseReferencedVBRuntime)' == ''
+                        and '$(VBRuntime)' == ''
+                        and '$(TargetFrameworkIdentifier)' == '.NETCoreApp'
+                        and '$(_TargetFrameworkVersionWithoutV)' >= '3.0'">
+    <UseReferencedVBRuntime>true</UseReferencedVBRuntime>
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(VBRuntime)' == ''
+                        and '$(UseReferencedVBRuntime)' != true
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.VisualBasic.targets]:
```suggestion
                        and '$(UseReferencedVBRuntime)' != 'true'
```

**Comment by nguerrera** [comment]
@livarcocc @dotnet/dotnet-cli Can I get some eyes on changes since review:

1. [force push](https://github.com/dotnet/sdk/compare/66df7f6bfab2f5553a81cb56c6c59a9d46353e73..fa763cc58a1cb90bf50e8e853528c673ed137537) to use TestProject instead of TestAsset, with the goal of making the test work x-plat by using Microsoft.NETFramework.ReferenceAssemblies.

2. 0e26306 to disable the net45 test when net45 targeting pack is not installed because of https://github.com/dotnet/core-sdk/issues/3228, which meant that (1) didn't actually work.

**Comment by nguerrera** [comment]
Sigh, one more adjustment for x-plat:  2523163

I'll wait for this to actually turn green before asking for a final review. :)

**Review by livarcocc — APPROVED** [review]

**Pull request merged**